vmctl: support of the remote read protocol (#3232)
VictoriaMetrics/VictoriaMetrics: app/vmctl/README.md

2020/02/23 15:51:07 Total time: 7.153158218s
```

## Migrating data by remote read protocol

> `vmctl` supports the `remote-read` mode for migrating data from any database which supports Prometheus remote read protocol 

Reviewer: ```suggestion
`vmctl` supports the `remote-read` mode for migrating data from any database which supports [Prometheus remote read API](https://prometheus.io/docs/prometheus/latest/querying/remote_read_api/) 
```

Author: fixed

to VictoriaMetrics time-series database.

See `./vmctl remote-read --help` for details and full list of flags.

> Migration is based on the remote reading and allows transparently fetch samples. 
> Remote read protocol supports two different modes `SAMPLES` and `STREAMED_XOR_CHUNKS`. You can read more about it 
> [here](https://prometheus.io/blog/2019/10/10/remote-read-meets-streaming/).

Reviewer: ```suggestion
```

Author: fixed

> Please notice at that moment not all TSDB support `STREAMED_XOR_CHUNKS` mode and can handle stream requests.

Reviewer: ```suggestion
_Please note, vmctl expects the source TSDB to support [STREAMED_XOR_CHUNKS](https://prometheus.io/docs/prometheus/latest/querying/remote_read_api/#streamed-chunks) mode for response streaming. [SAMPLES](https://prometheus.io/docs/prometheus/latest/querying/remote_read_api/#samples) mode is not supported due to its inefficiency. See more details [here](https://prometheus.io/blog/2019/10/10/remote-read-meets-streaming/)._
```

Author: fixed

To start the migration process some flags should be defined:
1. `--remote-read-src-addr` - address to perform read from;

> 2. `--vm-addr` - address for single-node VM is usually equal to `--httpListenAddr`, and for cluster version;

Reviewer 2: Does it append `/insert` for vminsert in cluster mode automatically?

Author: no, it doesn't user should define it by himself like `http://<vminsert>:8480/insert/<accountID>/prometheus`

Reviewer 2: Then I'd suggest adding a note about URL format here to make it easier to use

Author: note added

is equal to `--httpListenAddr` flag of vminsert component (for example `http://<vminsert>:8480/insert/<accountID>/prometheus`);

> 3. `--remote-read-filter-time-start` - the time filter in RFC3339 format to select timeseries with timestamp equal or higher than provided value. E.g. '2020-01-01T20:07:00Z';

Reviewer: does it have `time-end` filter?

Author: yes it does but this filter is not required, so I'm not describing it because the user can check it via 
`./vmctl remote-read --help`

Author: if user do not define it we use `time.Now()` for `time-end`

Reviewer: Ok. This needs to be outlined in docs.

> 4. `--remote-read-step-interval` - split export data into chunks. Valid values are `month, day, hour, minute`;

Reviewer: why would we need to support this for remote-read? What would be the benefit?

Reviewer: But people don't know that difference. All they want is just to migrate the data, and default settings should allow making that in an optimal way.
Is breaking-by-intervals beneficial for streaming mode? How does vmctl behave in non-stream mode when migrating gigabytes of data?

As soon as required flags are provided and all endpoints are accessible, `vmctl` will start the fetch process 

> from remote read API. Basically, it just fetches all timeseries from the provided database and sends import requests 

Reviewer: does it fetch time series one by one?

Author: I haven't implemented a concurrent fetch process. It can be the next step in the fetch process.
As I saw in my test logs series was fetched one by one.

for each timeseries and pass results to VM importer. VM importer then accumulates received samples in batches and 
sends import requests to VM.

The importing process example for local installation of Prometheus
and single-node VictoriaMetrics(`http://localhost:8428`):

```
./vmctl remote-read \
--remote-read-src-addr=http://127.0.0.1:9091 \
--remote-read-filter-time-start=2021-10-18T00:00:00Z \
--remote-read-step-interval=hour \
--vm-addr=http://127.0.0.1:8428 \

> --vm-concurrency=6

Reviewer: why do you need to specify concurrency in example?

Author: There was an issue with show statistics at the end of the process. 
When we add this flag part of the output hides the statistic when the process is finished.  
So I added this flag to see that the whole information is accessible to the user.

Reviewer: I don't get it( How concurrency affects the output?

Split defined times into 8798 ranges to import. Continue? [Y/n]
VM worker 0:↘ 127177 samples/s
VM worker 1:↘ 140137 samples/s
VM worker 2:↘ 151606 samples/s
VM worker 3:↘ 130765 samples/s
VM worker 4:↘ 131904 samples/s
VM worker 5:↘ 132693 samples/s
Processing ranges: 8798 / 8798 [█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████] 100.00%
2022/10/19 16:45:37 Import finished!
2022/10/19 16:45:37 VictoriaMetrics importer stats:

>   idle duration: 6m57.793987511s;

Reviewer: hm, idle duration is 7 times bigger than import duration. It means vmctl was mostly waiting for responses from the source. Do you know why so?

Author: This test was in a single Prometheus instance but this instance was a part of Thanos setup I think this is the first reason. Second, I think that system should collect all series for one hour and after that pack them into protobuf representation and send. After we should unpack it and prepare samples.

  time spent while importing: 1m18.463744801s;
  total samples: 25348208;
  samples/s: 323056.31;
  total bytes: 669.7 MB;
  bytes/s: 8.5 MB;
  import requests: 127;
  import requests retries: 0;
2022/10/19 16:45:37 Total time: 1m19.406283424s
```

### Filtering

The filtering consists of two parts: by label and label value, and time.

> Filtering by time may be configured via flags `--remote-read-filter-time-start` and `--remote-read-filter-time-end`

Reviewer: Is this true? I thought remote-read API does that, doesn't it?

Author: For getting data we should define the next struct
```golang
prompb.Query{
		StartTimestampMs: filter.StartTimestampMs,
		EndTimestampMs:   filter.EndTimestampMs - 1,
		Matchers:         []*prompb.LabelMatcher{m},
}
```

Where we must define the start and the end of collecting data

Reviewer: Oh, the comment is mis placed. It was related to the "This filter is applied twice: to drop blocks out of range and to filter timeseries in blocks with overlapping time ranges." part.

Author: removed this part of the code. you are right

in RFC3339 format. This filter is applied twice: to drop blocks out of range and to filter timeseries in blocks with
overlapping time ranges.

Also, you can add flags for filtering by label and label value via flags `--remote-read-filter-label` and `--remote-read-filter-label-value`.

> in string format. For example, you can provide `--remote-read-filter-label=__name__` and `--remote-read-filter-label-value=.*`, and

Reviewer: Not a good example. I assume, import will fetch all time series by default, even if filters are omitted. So in this example it is better to specify something which would actually filter the output.

Author: Yes it will fetch all time-series by default, because of the next code example
```golang
if filter.Label == "" && filter.LabelValue == "" {
		ms = labels.MustNewMatcher(labels.MatchRegexp, labels.MetricName, ".+")
	} else {
		ms, err = labels.NewMatcher(labels.MatchRegexp, filter.Label, filter.LabelValue)
		if err != nil {
			return nil, err
		}
	}
```
Ok, I will specify some other examples)

Author: example added and tested

all timeseries will be collected from provided source database.

> ### Configuration

Reviewer 2: Do we need this section here if there is just a reference to configuration flags?
I'd suggest either remove it or add configuration flags here.

Author: We have similar sections in other featuers

    vmctl prometheus --prom-snapshot thanos-data --vm-addr http://victoria-metrics:8428
    ```

### Remote read protocol

> At that moment Thanos only supports gRPC remote read protocol, but also they are [recommended](https://thanos.io/tip/thanos/integrations.md/#storeapi-as-prometheus-remote-read)

Reviewer: ```suggestion
Currently, Thanos doesn't support streaming remote read protocol. It is [recommended](https://thanos.io/tip/thanos/integrations.md/#storeapi-as-prometheus-remote-read)
```

to use [thanos-remote-read](https://github.com/G-Research/thanos-remote-read) a proxy, that allows exposing any Thanos 
service (or anything that exposes gRPC StoreAPI e.g. Querier) via Prometheus remote read protocol.

If you want to migrate data, you should run [thanos-remote-read](https://github.com/G-Research/thanos-remote-read) proxy
and define the Thanos store address `./thanos-remote-read -store 127.0.0.1:19194`. 
It is important to know that `store` flag is Thanos Store API gRPC endpoint. 

> Also, it is important to know that thanos-remote-read proxy  doesn't support `STREAMED_XOR_CHUNKS` mode.

Reviewer: if it doesn't support streaming, then migration via vmctl will hilariously expensive, I assume. wdyt?

Author: Yes it can't be expensive, but if we use filters 
`--remote-read-filter-time-start` and `--remote-read-filter-time-end` it helps with this problem

```
./vmctl remote-read \ 
--remote-read-src-addr=http://127.0.0.1:9009/prometheus \
--remote-read-filter-time-start=2021-10-18T00:00:00Z \
--remote-read-step-interval=hour \
--remote-read-src-check-alive=false \
--vm-addr=http://127.0.0.1:8428 \
--vm-concurrency=6 
```
And when the process finishes, you will see the following:

```
Split defined times into 8842 ranges to import. Continue? [Y/n]
VM worker 0:↗ 3863 samples/s
VM worker 1:↗ 2686 samples/s
VM worker 2:↗ 2620 samples/s
VM worker 3:↗ 2705 samples/s
VM worker 4:↗ 2643 samples/s
VM worker 5:↗ 2593 samples/s
Processing ranges: 8842 / 8842 [█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████] 100.00%
2022/10/21 12:09:49 Import finished!
2022/10/21 12:09:49 VictoriaMetrics importer stats:
  idle duration: 0s;
  time spent while importing: 3.82640757s;
  total samples: 160232;
  samples/s: 41875.31;
  total bytes: 11.3 MB;
  bytes/s: 3.0 MB;
  import requests: 6;
  import requests retries: 0;
2022/10/21 12:09:49 Total time: 4.71824253s
```

> It is important to know that if you run your Mimir installation in multi-tenant mode, remote read protocol

Reviewer 2: This paragraph references to Mimir, but it is in Cortex section, is it a typo or it shouldn't be here?

Author: Fixed

## Migrating data from Mimir

Mimir has similar implemintation as Cortex and also support of the Prometheus remote read protocol. That means
`vmctl` in mode `remote-read` may also be used for Mimir historical data migration.
These instructions may vary based on the details of your Mimir configuration.
Please read carefully and verify as you go.

### Remote read protocol

If you want to migrate data, you should check your Mimir configuration in the section
```yaml
api:
  prometheus_http_prefix:
```

If you defined some prometheus prefix, you should use it when you define flag `--remote-read-src-addr=http://127.0.0.1:9009/{prometheus_http_prefix}`.
By default, Mimir uses the `prometheus` path prefix, so you should define the flag `--remote-read-src-addr=http://127.0.0.1:9009/prometheus`.
Mimir API doesn't support Prometheus `health` API, so we should disable the health check when we start the migration.
It is possible todo by setting flag as `--remote-read-src-check-alive=false`

Mimir supports both remote read mode, so you can use `STREAMED_XOR_CHUNKS` mode and `SAMPLES` mode.
When you run Mimir, it exposes a port to serve HTTP on `8080 by default`.

> Next example of the local installation was in multi-tenant mode (3 instances of mimir) with nginx as load balancer.

Reviewer 2: It seems to me like we should remove nginx LB details from doc and just use "Mimir API endpoint" which should be self-explanatory for users.
Users might think that nginx is required for this to work when it doesn't.

Reviewer 2: I think it will be better to remove note regarding LB completely as user might still think that some kind of proxy is required, while we only need access to Mimir endpoint.

Author: I want to show the user that he can run it in multitenant mode and how to use it when you want to get data from that setup. I can describe that nginx is not necessary for the setup. But in that example user can know that he can use single port or use proxy if it is needed
